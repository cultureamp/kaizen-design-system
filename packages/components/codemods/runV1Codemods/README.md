# runV1Codemods

A comprehensive codemod that runs all V1 migration codemods in the ideal sequence.

## Overview

This codemod executes all 17 V1 migration codemods in the order as pre-work to migrate to `kaizen/component@2.0.0` Kaizen components. After this you will still need to run the `migrateV2NextToCurrent` and `renameV2ComponentImportsAndUsages` after upgrading to `kaizen/component@2.0.0`

This codemod runner handles the following:

- Icon upgrades (`upgradeIconV1`)
- Button migrations (`upgradeV1Buttons`)
- GuidanceBlock actions migration (`migrateGuidanceBlockActionsToActionsSlot`)
- Mood to variant migrations for multiple components
- Variant to color migrations
- Type to variant migrations for notifications
- Removal of deprecated props

## Usage

```bash
pnpm kaizen-codemod {target-directory} runV1Codemods
```

## Example

```bash
pnpm kaizen-codemod src/components/MockComponent runV1Codemods
```

## What it does

The codemod runs the following transformations in sequence:

1. **upgradeIconV1** - Migrates V1 Icon components to next Icon
2. **upgradeV1Buttons** - Migrates V1 Button and IconButton to next Button/LinkButton
3. **migrateGuidanceBlockActionsToActionsSlot** - Migrates GuidanceBlock actions prop to actionsSlot
4. **migrateBrandMomentMoodToVariant** - Changes `mood` prop to `variant` on BrandMoment
5. **migrateCardVariantToColor** - Changes `variant` prop to `color` on Card
6. **migrateConfirmationModalMoodsToVariant** - Changes `mood` prop to `variant` on ConfirmationModal
7. **migrateEmptyStateIllustrationTypeToVariant** - Changes `illustrationType` prop to `variant` on EmptyState
8. **migrateGlobalNotificationTypeToVariant** - Changes `type` prop to `variant` on GlobalNotification
9. **migrateInformationTileMoodToVariant** - Changes `mood` prop to `variant` on InformationTile
10. **migrateInlineNotificationTypeToVariant** - Changes `type` prop to `variant` on InlineNotification
11. **migrateMultiActionTileMoodToVariant** - Changes `mood` prop to `variant` on MultiActionTile
12. **migrateNotificationTypeToVariant** - Changes `type` prop to `variant` on Notification
13. **migrateProgressBarMoodToColor** - Changes `mood` prop to `color` on ProgressBar
14. **migrateToastNotificationTypeToVariant** - Changes `type` prop to `variant` on ToastNotification
15. **migrateWellVariantToColor** - Changes `variant` prop to `color` on Well
16. **removeInputEditModalMood** - Removes deprecated `mood` prop from InputEditModal
17. **removePopoverVariant** - Removes deprecated `variant` and `customIcon` props from Popover

## Error Handling

- If any individual codemod fails, the process continues with the remaining codemods
- Detailed logging shows the start and completion status of each codemod
- A summary is provided at the end showing successful vs failed codemods

## Cautions and caveats

### Reducing the blast radius of the codemods

This running will create a lot of diffs and changes to your components as the traverses the entire directory given.

To reduce the scope, consider either running this on component or page directories or running the mode incrementally, ie:

`pnpm kaizen-codemod src/pages/MockPage runV1Codemods`

This can make it easier to review and merge in changes with confidence.

### Prettier diffs

As the codemod will rewrites files, this can cause arbitrary changes that will be resolved by `prettier` formatting. We recommend running the following to reduce the noise in your diffs:

`pnpm prettier --write "**/*.tsx"`
