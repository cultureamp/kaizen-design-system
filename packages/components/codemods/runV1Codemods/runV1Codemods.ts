import fs from 'fs'
import { transformBrandMomentMoodToVariant } from '../migrateBrandMomentMoodToVariant/transformBrandMomentMoodToVariant'
import { transformCardVariantToColor } from '../migrateCardVariantToColor/transformCardVariantToColor'
import { transformConfirmationModalMoodsToVariant } from '../migrateConfirmationModalMoodsToVariant/transformConfirmationModalMoodsToVariant'
import { transformEmptyStateIllustrationTypeToVariant } from '../migrateEmptyStateIllustrationTypeToVariant/transformEmptyStateIllustrationTypeToVariant'
import { migrateGuidanceBlockActionsToActionsSlot } from '../migrateGuidanceBlockActionsToActionsSlot/migrateGuidanceBlockActionsToActionsSlot'
import { transformInformationTileMoodToVariant } from '../migrateInformationTileMoodToVariant/transformInformationTileMoodToVariant'
import { transformMultiActionTileMoodToVariant } from '../migrateMultiActionTileMoodToVariant/transformMultiActionTileMoodToVariant'
import { transformNotificationTypeToVariant } from '../migrateNotificationTypeToVariant/migrateNotificationTypeToVariant'
import { transformProgressBarMoodToColor } from '../migrateProgressBarMoodToColor/transformProgressBarMoodToColor'
import { transformWellVariantToColor } from '../migrateWellVariantToColor/transformWellVariantToColor'
import { removeInputEditModalMood } from '../removeInputEditModalMood/removeInputEditModalMood'
import { removePopoverVariant } from '../removePopoverVariant/removePopoverVariant'
import { upgradeIconV1 } from '../upgradeIconV1/upgradeIconV1'
import { upgradeV1Buttons } from '../upgradeV1Buttons/upgradeV1Buttons'
import { transformComponentsInDir, transformComponentsInDirByPattern } from '../utils'

interface CodemodConfig {
  name: string
  runner: (targetDir: string) => void
}

const codemods: CodemodConfig[] = [
  {
    name: 'upgradeIconV1',
    runner: (dir) => {
      transformComponentsInDirByPattern(dir, 'Icon$', (tagNames) => [upgradeIconV1(tagNames)])
    },
  },
  {
    name: 'upgradeV1Buttons',
    runner: (dir) => {
      transformComponentsInDir(dir, ['IconButton', 'Button'], (tagNames) => [
        upgradeV1Buttons(tagNames),
      ])
    },
  },
  {
    name: 'migrateGuidanceBlockActionsToActionsSlot',
    runner: (dir) => {
      transformComponentsInDir(dir, ['GuidanceBlock'], (tagNames) => [
        migrateGuidanceBlockActionsToActionsSlot(tagNames),
      ])
    },
  },
  {
    name: 'migrateBrandMomentMoodToVariant',
    runner: (dir) => {
      transformComponentsInDir(dir, ['BrandMoment'], (tagNames) => [
        transformBrandMomentMoodToVariant(tagNames),
      ])
    },
  },
  {
    name: 'migrateCardVariantToColor',
    runner: (dir) => {
      transformComponentsInDir(dir, ['Card'], (tagNames) => [transformCardVariantToColor(tagNames)])
    },
  },
  {
    name: 'migrateConfirmationModalMoodsToVariant',
    runner: (dir) => {
      transformComponentsInDir(dir, ['ConfirmationModal'], (tagNames) => [
        transformConfirmationModalMoodsToVariant(tagNames),
      ])
    },
  },
  {
    name: 'migrateEmptyStateIllustrationTypeToVariant',
    runner: (dir) => {
      transformComponentsInDir(dir, ['EmptyState'], (tagNames) => [
        transformEmptyStateIllustrationTypeToVariant(tagNames),
      ])
    },
  },
  {
    name: 'migrateGlobalNotificationTypeToVariant',
    runner: (dir) => {
      transformComponentsInDir(dir, ['GlobalNotification'], (tagNames) => [
        transformNotificationTypeToVariant(tagNames),
      ])
    },
  },
  {
    name: 'migrateInformationTileMoodToVariant',
    runner: (dir) => {
      transformComponentsInDir(dir, ['InformationTile'], (tagNames) => [
        transformInformationTileMoodToVariant(tagNames),
      ])
    },
  },
  {
    name: 'migrateInlineNotificationTypeToVariant',
    runner: (dir) => {
      transformComponentsInDir(dir, ['InlineNotification'], (tagNames) => [
        transformNotificationTypeToVariant(tagNames),
      ])
    },
  },
  {
    name: 'migrateMultiActionTileMoodToVariant',
    runner: (dir) => {
      transformComponentsInDir(dir, ['MultiActionTile'], (tagNames) => [
        transformMultiActionTileMoodToVariant(tagNames),
      ])
    },
  },
  {
    name: 'migrateNotificationTypeToVariant',
    runner: (dir) => {
      transformComponentsInDir(dir, ['Notification'], (tagNames) => [
        transformNotificationTypeToVariant(tagNames),
      ])
    },
  },
  {
    name: 'migrateProgressBarMoodToColor',
    runner: (dir) => {
      transformComponentsInDir(dir, ['ProgressBar'], (tagNames) => [
        transformProgressBarMoodToColor(tagNames),
      ])
    },
  },
  {
    name: 'migrateToastNotificationTypeToVariant',
    runner: (dir) => {
      transformComponentsInDir(dir, ['ToastNotification'], (tagNames) => [
        transformNotificationTypeToVariant(tagNames),
      ])
    },
  },
  {
    name: 'migrateWellVariantToColor',
    runner: (dir) => {
      transformComponentsInDir(dir, ['Well'], (tagNames) => [transformWellVariantToColor(tagNames)])
    },
  },
  {
    name: 'removeInputEditModalMood',
    runner: (dir) => {
      transformComponentsInDir(dir, ['InputEditModal'], (tagNames) => [
        removeInputEditModalMood(tagNames),
      ])
    },
  },
  {
    name: 'removePopoverVariant',
    runner: (dir) => {
      transformComponentsInDir(dir, ['Popover'], (tagNames) => [removePopoverVariant(tagNames)])
    },
  },
]

export const runV1Codemods = (targetDir: string): void => {
  console.log(`📝 Running ${codemods.length} codemods on directory: ${targetDir}`)
  console.log('')

  if (fs.existsSync(targetDir) === false) {
    console.error(`Error: Target directory "${targetDir}" does not exist`)
    process.exit(1)
  }

  let successCount = 0

  for (const codemod of codemods) {
    try {
      console.log(`🔄 Starting codemod: ${codemod.name}`)
      codemod.runner(targetDir)
      console.log(`✅ Completed codemod: ${codemod.name}`)
      successCount++
    } catch (error) {
      console.log(`💥 Error in codemod: ${codemod.name} - exiting early

✅ ${successCount} codemods completed successfully
😢 ${codemods.length - successCount} codemods did not run
Address the issues in the following codemod before proceeding:
        `)

      console.error(`${codemod.name}\n`)
      console.error(error)
      process.exit(1)
    }
    console.log('')
  }

  console.log('\n🎉 All codemods completed successfully!')
}
