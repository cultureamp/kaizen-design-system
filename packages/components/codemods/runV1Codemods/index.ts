import { runV1Codemods } from './runV1Codemods'

const run = async (): Promise<void> => {
  console.log(`ðŸš€ Running all V1 codemods in sequence:
  1. upgradeIconV1
  2. upgradeV1Buttons
  3. migrateGuidanceBlockActionsToActionsSlot
  4. migrateBrandMomentMoodToVariant
  5. migrateCardVariantToColor
  6. migrateConfirmationModalMoodsToVariant
  7. migrateEmptyStateIllustrationTypeToVariant
  8. migrateGlobalNotificationTypeToVariant
  9. migrateInformationTileMoodToVariant
  10. migrateInlineNotificationTypeToVariant
  11. migrateMultiActionTileMoodToVariant
  12. migrateNotificationTypeToVariant
  13. migrateProgressBarMoodToColor
  14. migrateToastNotificationTypeToVariant
  15. migrateWellVariantToColor
  16. removeInputEditModalMood
  17. removePopoverVariant
  `)

  const targetDir = process.argv[2]

  if (!targetDir) {
    console.error('Error: Target directory is required')
    process.exit(1)
  }

  await runV1Codemods(targetDir)
}

run().catch((error) => {
  console.error('Fatal error running codemods:', error)
  process.exit(1)
})
