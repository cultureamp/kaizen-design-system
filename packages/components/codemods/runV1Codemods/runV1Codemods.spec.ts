import fs from 'fs'
import os from 'os'
import path from 'path'
import { runV1Codemods } from './runV1Codemods'

const cleanupTestDirectory = (tempDir: string): void => {
  if (fs.existsSync(tempDir)) {
    fs.rmSync(tempDir, { recursive: true, force: true })
  }
}

// Helper to read files from test directory
const readTestFiles = (tempDir: string): Record<string, string> => {
  const files: Record<string, string> = {}

  const readDir = (dirPath: string, basePath = ''): void => {
    const items = fs.readdirSync(dirPath)

    items.forEach((item) => {
      const fullPath = path.join(dirPath, item)
      const relativePath = path.join(basePath, item)

      if (fs.statSync(fullPath).isDirectory()) {
        readDir(fullPath, relativePath)
      } else if (item.endsWith('.tsx') || item.endsWith('.ts')) {
        files[relativePath] = fs.readFileSync(fullPath, 'utf8')
      }
    })
  }

  readDir(tempDir)
  return files
}

const copyDirectoryToTemp = (sourceDir: string): string => {
  const tempDir = fs.mkdtempSync(path.join(os.tmpdir(), 'runV1Codemods-snapshot-'))

  const copyRecursive = (src: string, dest: string): void => {
    const stat = fs.statSync(src)

    if (stat.isDirectory()) {
      if (!fs.existsSync(dest)) {
        fs.mkdirSync(dest, { recursive: true })
      }
      const items = fs.readdirSync(src)
      items.forEach((item) => {
        copyRecursive(path.join(src, item), path.join(dest, item))
      })
    } else {
      fs.copyFileSync(src, dest)
    }
  }

  copyRecursive(sourceDir, tempDir)
  return tempDir
}

describe('runV1Codemods', () => {
  describe('MockComponentDir Snapshot Tests', () => {
    it('should transform MockComponentDir files and match snapshot', async () => {
      const mockSourceDir = path.resolve(__dirname, '__fixtures__', 'MockComponentDir')
      const tempDir = copyDirectoryToTemp(mockSourceDir)

      try {
        await runV1Codemods(tempDir)
        const transformedFiles = readTestFiles(tempDir)

        // Snapshot test - this will create/update snapshots on first run
        expect(transformedFiles).toMatchSnapshot()

        // Specific assertions to ensure transformations occurred
        expect(transformedFiles['MockComponent.tsx']).toContain('actionsSlot')
        expect(transformedFiles['MockComponent.tsx']).not.toContain('actions={{')
        expect(transformedFiles['MockComponent.tsx']).toContain(
          'import { Button } from "@kaizen/components/next"',
        )

        // Check that subcomponent was transformed
        expect(transformedFiles['subcomponents/MockSubcomponent.tsx']).toContain('onPress')
        expect(transformedFiles['subcomponents/MockSubcomponent.tsx']).toContain(
          'import { Button } from "@kaizen/components/next"',
        )
      } finally {
        cleanupTestDirectory(tempDir)
      }
    })
  })
})
