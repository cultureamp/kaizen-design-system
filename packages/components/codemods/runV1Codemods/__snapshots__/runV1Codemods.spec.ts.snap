// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`runV1Codemods > MockComponentDir Snapshot Tests > should transform MockComponentDir files and match snapshot 1`] = `
{
  "MockComponent.tsx": "// @ts-nocheck this is a mock file - we can ignore this
import React from 'react';
import { BrandMoment, Card, GlobalNotification, GuidanceBlock, Informative, ProgressBar, Well, LinkButton } from '@kaizen/components';
import { Button } from "@kaizen/components/next";
import { MockSubcomponent } from './subcomponents/MockSubcomponent';

export const MockComponent = (): JSX.Element => {
    return (<div>
      <GuidanceBlock illustration={<Informative alt=""/>} text={{
            title: 'This is the call to action title',
            description: 'Mussum Ipsum, cacilds vidis litro abertis. Suco de cevadiss, Ã© um leite divinis.',
        }} actionsSlot={<><Button onPress={() => undefined} variant="secondary" size="large">Action!</Button><LinkButton href='/cancel' variant="tertiary" size="large">Cancel</LinkButton></>}/>
      <BrandMoment variant="success">
        <div>Success message</div>
      </BrandMoment>
      <Card color="blue">
        <div>Card content</div>
      </Card>
      <ProgressBar color="green" value={75}/>
      <Well color="gray">
        <div>Well content</div>
      </Well>
      <GlobalNotification persistent={true} variant="success">
        Global notification content
      </GlobalNotification>
      <div className="additional-content">
        This is some additional content that is not part of the GuidanceBlock.
        <MockSubcomponent />
      </div>
    </div>);
};




 
",
  "index.ts": "export * from './'
",
  "subcomponents/MockSubcomponent.tsx": "import { Button } from "@kaizen/components/next";
import { LinkButton } from "@kaizen/components";
import React from 'react';

export const MockSubcomponent = (): JSX.Element => {
    return (<div>
      <Button onPress={() => undefined} variant="secondary" size="large">Mock Subcomponent Button</Button>
      <LinkButton href="#link" variant="secondary" size="large">Mock Subcomponent Link</LinkButton>
    </div>);
};
",
}
`;
